### PR TITLE
[Yaml] Fixed invalid Parser behavior

### LINKED ISSUE #27874: [Yaml] parses misplaced keys in strings
**Symfony version(s) affected**: 3.x/4.x

**Description**  

When you use a mapping key in a multi-line string, that value is mistakenly treated as part of the string while it should lead to a parser error instead.

**How to reproduce**  

```php
use Symfony\Component\Yaml\Yaml;

$yaml = <<<EOT
doctrine:
    dbal:yeah
        default_connection: monolith
EOT;

print_r(Yaml::parse($yaml));
```

ends up in this result:

```
Array
(
    [doctrine] => dbal:yeah default_connection: monolith
)
```

### PR DESCRIPTION
**Pull request by guiguiboy**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | yes
| New feature?  | no 
| BC breaks?    | no  
| Deprecations? | no
| Tests pass?   | yes 
| Fixed tickets | #27874 
| License       | MIT
| Doc PR        | NA

This fixes #27874
I'm not sure about the update in composer.json though. It seems a good idea because I was able to run composer update without the zip extension. If required, I'll remove it.

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -85,6 +85,7 @@
         "symfony/yaml": "self.version"
     },
     "require-dev": {
+        "ext-zip": "*",
```
Comment (fabpot) [inline composer.json]: This should be reverted as it does not seem to be related to the fix.

**Review by fabpot — DISMISSED** [review]
/cc @xabbuh

File: composer.json (hunk 1 above)
Comment (guiguiboy) [reply composer.json]: @fabpot Should I create a separate PR for this ?

Comment (xabbuh) [reply composer.json]: I don't think this change is valid at all. IIRC we need the extension for some parts.

Comment (stof) [reply composer.json]: for which part ? AFAIK, we don't need ext-zip to run tests (and then, we should mark the relevant tests as skipped instead of forbidding to install the dev environment).
And if this is about running composer, it does not require ext-zip. It requires having *either* ext-zip or the `unzip` command (i.e. having *something* to unzip). And it would still not be a requirement of Symfony.

File: src/Symfony/Component/Yaml/Parser.php [hunk 2]
```diff
@@ -381,6 +381,11 @@ private function doParse(string $value, int $flags)
                         if (0 === $this->offset && !$deprecatedUsage && isset($line[0]) && ' ' === $line[0]) {
                             throw new ParseException('Unable to parse.', $this->getRealCurrentLineNb() + 1, $this->currentLine, $this->filename);
                         }
+
+                        if ($this->offset >= 1 && preg_match('/:/', $this->currentLine) && $this->isNextLineIndented()) {
+                            throw new ParseException('Indented blocks after definition are not allowed', $this->currentLineNb + 1, $this->currentLine, $this->filename);
```
Comment (stof) [inline src/Symfony/Component/Yaml/Parser.php]: this needs to go through a deprecation process instead, as done for our previous changes making the parser stricter regarding the spec compliance.

**Review by stof — DISMISSED** [review]

File: composer.json (hunk 1 above)
Comment (xabbuh) [reply composer.json]: Our `simple-phpunit` script uses the zip extension to extract the downloaded PHPUnit archive.

Comment (guiguiboy) [reply composer.json]: The problem I had was that I cloned the repository and ran the tests directly without having the zip extension installed. This gave me some errors. After install the php-zip extension everything worked fine. That's why I also proposed this change. But as it's not related to the ticket this PR is linked, I'll revert this.

File: src/Symfony/Component/Yaml/Parser.php (hunk 2 above)
Comment (guiguiboy) [reply src/Symfony/Component/Yaml/Parser.php]: @stof Do you have an example I can follow?

Comment (xabbuh) [reply src/Symfony/Component/Yaml/Parser.php]: you can take a look at #23288 for example

File: src/Symfony/Component/Yaml/Parser.php [hunk 3]
```diff
@@ -381,6 +381,11 @@ private function doParse(string $value, int $flags)
                         if (0 === $this->offset && !$deprecatedUsage && isset($line[0]) && ' ' === $line[0]) {
                             throw new ParseException('Unable to parse.', $this->getRealCurrentLineNb() + 1, $this->currentLine, $this->filename);
                         }
+
+                        if ($this->offset >= 1 && preg_match('/:/', $this->currentLine) && $this->isNextLineIndented()) {
```
Comment (xabbuh) [inline src/Symfony/Component/Yaml/Parser.php]: we should use `strpos()` instead of `preg_match()` here

Comment (xabbuh) [reply src/Symfony/Component/Yaml/Parser.php]:
```yaml
data:
    dbal:wrong
        default_connection:monolith
```

Would trigger the deprecation too, wouldn't it?

Comment (xabbuh) [reply src/Symfony/Component/Yaml/Parser.php]: IMO that's not right. The first example is valid YAML. But I guess we could fix that by changing this to `false !== strpos($this->currentLine, ': ')`.

Comment (guiguiboy) [reply src/Symfony/Component/Yaml/Parser.php]: How can we check that ? Looking at the YAML 1.2 specification, I see 
"
Each node must be indented further than its parent node. All sibling nodes must use the exact same indentation level. However the content of each sibling node may be further indented independently.
"

Comment (xabbuh) [reply src/Symfony/Component/Yaml/Parser.php]: What do you mean with what to check? I am talking here about the colon not being followed by a space.

Comment (guiguiboy) [reply src/Symfony/Component/Yaml/Parser.php]: Are you sure the first example is also a valid YAML? Quoting from the specification : 
"Normally, YAML insists the “:” mapping value indicator be separated from the value by white space."
IMO it should also trigger the deprecation notice.

Comment (guiguiboy) [reply src/Symfony/Component/Yaml/Parser.php]: I replaced preg_match with strpos.

Comment (xabbuh) [reply src/Symfony/Component/Yaml/Parser.php]: Yes, in that case the last two lines are not mappings, but form a string that crosses multiple lines. This string in turn is the value of the `data` mapping key.

Comment (guiguiboy) [reply src/Symfony/Component/Yaml/Parser.php]: Ok I updated the check in order to trigger deprecation notice only on second scenario. I also updated the multiline test to make sure the value will still be correctly generated with a colon in the string.

File: src/Symfony/Component/Yaml/Tests/ParserTest.php [hunk 4]
```diff
@@ -735,9 +749,10 @@ public function testMultiLineStringLastResortParsing()
   You can have things that don't look like strings here
   true
   yes you can
+  with a: colon
```
Comment (xabbuh) [inline src/Symfony/Component/Yaml/Tests/ParserTest.php]: This does not seem to be allowed. Does it not trigger the deprecation?

Comment (guiguiboy) [reply src/Symfony/Component/Yaml/Tests/ParserTest.php]: No it doesn't here because in the check I also check if the line is indented

Comment (xabbuh) [reply src/Symfony/Component/Yaml/Tests/ParserTest.php]: Then the indentation handling looks wrong to me. Afaik this is not valid YAML.

Comment (guiguiboy) [reply src/Symfony/Component/Yaml/Tests/ParserTest.php]: I'll simplify the check. If we are dealing with a multiline string, whenever we encounter ': ' a deprecation notice would be thrown.

**Comment by fabpot** [comment]
@xabbuh Can you review this one one more time please?

**Comment by stof** [comment]
As this introduce a new deprecation, it should go in master, not in 4.1. We don't introduce deprecations in patch releases.

**Comment by guiguiboy** [comment]
Hello, I missed your message. I'll rebase shortly.

**Comment by xabbuh** [comment]
@guiguiboy I will look into the changes in more depth shortly. Thanks so far. 👍 Can you in the meantime please document the deprecation in the upgrade files and in the changelog file of the Yaml component?

**Comment by guiguiboy** [comment]
I updated the files supposing it will go in version 5.0.0. If required, I can change back to 4.2.0.

File: src/Symfony/Component/Yaml/CHANGELOG.md [hunk 5]
```diff
@@ -1,6 +1,12 @@
 CHANGELOG
 =========
 
+5.0.0
+-----
+
+ * The parser is now stricter and will trigger a deprecation notice when a
```
Comment (xabbuh) [inline src/Symfony/Component/Yaml/CHANGELOG.md]: I would reword this paragraph like this:

> Using a mapping inside a multi-line string is deprecated and will throw a `ParseException` in 5.0.

**Review by xabbuh — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Yaml/Parser.php [hunk 6]
```diff
@@ -381,6 +381,11 @@ private function doParse(string $value, int $flags)
                         if (0 === $this->offset && !$deprecatedUsage && isset($line[0]) && ' ' === $line[0]) {
                             throw new ParseException('Unable to parse.', $this->getRealCurrentLineNb() + 1, $this->currentLine, $this->filename);
                         }
+
+                        if ($this->offset >= 1 && strpos($line, ': ')) {
```
[the remark this answers is not in the document]
Comment (guiguiboy) [reply src/Symfony/Component/Yaml/Parser.php]: Done. For the offset >= 1 condition, as far as I remember, I added it because it broke some tests. As of now, without this condition, the test passes.

File: src/Symfony/Component/Yaml/Parser.php [hunk 7]
```diff
@@ -381,6 +381,11 @@ private function doParse(string $value, int $flags)
...
+                            @trigger_error('Support for mapping keys in multiline blocks is deprecated since version 4.1.', E_USER_DEPRECATED);
```
[the remark this answers is not in the document]
Comment (guiguiboy) [reply src/Symfony/Component/Yaml/Parser.php]: Done

File: UPGRADE-5.0.md [hunk 8]
```diff
@@ -272,3 +272,9 @@ Workflow
  * `add` method has been removed use `addWorkflow` method in `Workflow\Registry` instead.
  * `SupportStrategyInterface` has been removed, use `WorkflowSupportStrategyInterface` instead.
  * `ClassInstanceSupportStrategy` has been removed, use `InstanceOfSupportStrategy` instead.
+
+Yaml
+----
+
+ * The parser is now stricter and will trigger a deprecation notice when a
```
[the remark this answers is not in the document]
Comment (guiguiboy) [reply UPGRADE-5.0.md]: Done

File: src/Symfony/Component/Yaml/CHANGELOG.md [hunk 9]
```diff
@@ -1,6 +1,12 @@
 CHANGELOG
 =========
 
+5.0.0
```
[the remark this answers is not in the document]
Comment (guiguiboy) [reply src/Symfony/Component/Yaml/CHANGELOG.md]: Done

File: src/Symfony/Component/Yaml/CHANGELOG.md (hunk 5 above)
Comment (guiguiboy) [reply src/Symfony/Component/Yaml/CHANGELOG.md]: Done

**Review by xabbuh — APPROVED** [review]
with a minor tweak to the deprecation message

**Comment by xabbuh** [comment]
Thank you @guiguiboy.
